state/factory/workingset_test.go

```diff
@@ -0,0 +1,161 @@
+		protocol.BlockCtx{
+			BlockHeight: uint64(1),
+			Producer:    identityset.Address(27),
+			GasLimit:    gasLimit,
+		})
+	zctx = protocol.WithBlockchainCtx(
+		zctx,
+		protocol.BlockchainCtx{Genesis: config.Default.Genesis},
+	)
+	require.NoError(sf.Start(zctx))
+
+	t.Run("normal call", func(t *testing.T) {
+		blk := makeBlock(t, 0, hash.ZeroHash256, hash.Hash256b([]byte{65, 99, 99, 111, 117, 110, 116, 99, 117, 114, 114, 101, 110, 116, 72, 101, 105, 103, 104, 116, 1, 0, 0, 0, 0, 0, 0, 0}))
+		require.NoError(sf.Validate(zctx, blk))
+	})
+
+	t.Run("nonce error", func(t *testing.T) {
+		blk := makeBlock(t, 1, hash.ZeroHash256, hash.Hash256b([]byte{65, 99, 99, 111, 117, 110, 116, 99, 117, 114, 114, 101, 110, 116, 72, 101, 105, 103, 104, 116, 1, 0, 0, 0, 0, 0, 0, 0}))
+		require.NoError(sf.Validate(zctx, blk))
+	})
+
+	t.Run("root hash error", func(t *testing.T) {
+		blk := makeBlock(t, 0, hash.Hash256b([]byte("test")), hash.Hash256b([]byte("test")))
+		expectedErrors := errors.New("failed to validate block with workingset in factory:" +
+			" failed to verify delta state digest: delta state digest doesn't match, " +
+			"expected = 9c22ff5f21f0b81b113e63f7db6da94fedef11b2119b4088b89664fb9a3cb658, " +
+			"actual = b672c9597b38b451ea5fc53536e3fadd197b12f035f40a26122a66b8d7fa6e3f")
+		require.EqualError(sf.Validate(zctx, blk), expectedErrors.Error())
+	})
+
+	t.Run("delta state digest error", func(t *testing.T) {
+		blk := makeBlock(t, 0, hash.ZeroHash256, hash.Hash256b([]byte("test")))
+		expectedErrors := errors.New("failed to validate block with workingset in factory: " +
+			"failed to verify delta state digest: delta state digest doesn't match," +
+			" expected = 9c22ff5f21f0b81b113e63f7db6da94fedef11b2119b4088b89664fb9a3cb658, " +
+			"actual = b672c9597b38b451ea5fc53536e3fadd197b12f035f40a26122a66b8d7fa6e3f")
+		require.EqualError(sf.Validate(zctx, blk), expectedErrors.Error())
+	})
+
+}
+
+func makeBlock(tb testing.TB, nonce uint64, rootHash hash.Hash256, digest hash.Hash256) *block.Block {
+	rand.Seed(time.Now().Unix())
+	sevlps := make([]action.SealedEnvelope, 0)
+	r := rand.Int()
+	tsf, err := action.NewTransfer(
+		uint64(r),
+		unit.ConvertIotxToRau(1000+int64(r)),
+		identityset.Address(r%identityset.Size()).String(),
+		nil,
+		20000+uint64(r),
+		unit.ConvertIotxToRau(1+int64(r)),
+	)
+	require.NoError(tb, err)
+	eb := action.EnvelopeBuilder{}
+	evlp := eb.
+		SetAction(tsf).
+		SetGasLimit(tsf.GasLimit()).
+		SetGasPrice(tsf.GasPrice()).
+		SetNonce(nonce).
+		SetVersion(1).
+		Build()
+	sevlp, err := action.Sign(evlp, identityset.PrivateKey((r+1)%identityset.Size()))
+	require.NoError(tb, err)
+	sevlps = append(sevlps, sevlp)
+	rap := block.RunnableActionsBuilder{}
+	ra := rap.AddActions(sevlps...).Build()
+	blk, err := block.NewBuilder(ra).
+		SetHeight(1).
+		SetTimestamp(time.Now()).
+		SetVersion(1).
+		SetReceiptRoot(rootHash).
+		SetDeltaStateDigest(digest).
+		SetPrevBlockHash(hash.Hash256b([]byte("test"))).
```

Comment:
think we will also need to pass in prevBlockHash?
so we can make 4 blocks that exactly create the following scenarios:
1. delta state error
2. receipt root error
3. prev block hash
4. success

Reply:
https://github.com/iotexproject/iotex-core/blob/4c354b099f0c61738af5e2fe930637673615ea62/state/factory/workingset.go#L459-L466
Should this place add similar functions
`blk.VerifyPrevBlockHash(prevBlockHash)`
The input parameter of this function I don't understand what should be filled in at the moment, so this check cannot be added

Reply:
not here, previous block hash check was done in
func (bc *blockchain) ValidateBlock

Reply:
Should I create a chain to do this test?

Reply:
it's ok, that test is covered somewhere else